lib/utils/typical/parser.go

```diff
@@ -249,9 +269,46 @@ func getProperty[TEnv any](mapVal, keyVal any) (any, error) {
 	if dynamicMap, ok := mapVal.(indexExpressionBuilder[TEnv]); ok {
 		return dynamicMap.buildIndexExpression(keyExpr), nil
 	}
+
+	// Only allow falling back to an untyped expression if the parser was constructed
+	// to allow unknown identifiers. This ensures compile time type safety for all
+	// parsers that don't explicitly opt in to the more dangerous behavior required to
+	// support dynamic fields.
+	if p.spec.GetUnknownIdentifier != nil {
+		if mapExpr, ok := mapVal.(Expression[TEnv, any]); ok {
+			return untypedMapExpr[TEnv]{mapExpr, keyExpr}, nil
+		}
+	}
+
 	return nil, trace.Wrap(unexpectedTypeError[map[string]string](mapVal), "cannot take index of unexpected type")
 }
 
+type untypedMapExpr[TEnv any] struct {
```

Comment:
nit: (I know I'm the one that wrote this while we were chatting about it but) this is essentially the same as `propertyExpr` below, just `mapExpr` doesn't have a known result type. Can you rename it to `untypedPropertyExpr` and move the definition below `propertyExpr`'s?